codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -600,7 +604,48 @@ protected void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Htt
      */
     protected void onStreamError(ChannelHandlerContext ctx, @SuppressWarnings("unused") Throwable cause,
                                  StreamException http2Ex) {
-        resetStream(ctx, http2Ex.streamId(), http2Ex.error().code(), ctx.newPromise());
+        int streamId = http2Ex.streamId();
+
+        //if this is caused by reading headers that are too large, send a header with status 431
+        if (http2Ex instanceof Http2Exception.HeaderListSizeException &&
+            ((Http2Exception.HeaderListSizeException) http2Ex).duringDecode() &&
+            connection().isServer()) {
+
+            final Http2Stream stream = connection().stream(streamId);
+
+            // NOTE We have to check to make sure that a stream exists before we send our reply.
+            // We likely always create the stream below as the stream isn't created until the
+            // header block is completely processed.
+
+            // The case of a streamId referring to a stream which was already closed is handled
+            // by createStream and will land us in the catch block below
+            if (stream == null) {
+                try {
+                    encoder.connection().remote().createStream(streamId, true);
+                } catch (Http2Exception e) {
+                    resetUnknownStream(ctx, streamId, http2Ex.error().code(), ctx.newPromise());
+                    return;
+                }
+            }
+
+            // ensure that we have not already sent headers on this stream
+            if (!stream.isHeadersSent()) {
+                handleServerHeaderDecodeSizeError(ctx, streamId);
+            }
```

Comment:
since we already have a `Http2Stream` object consider extracting out the portions of `resetStream` into a function which takes the `Http2Stream` instead of the streamId to avoid the duplicate map lookup:

```java
private ChannelFuture resetStream(final ChannelHandlerContext ctx, Http2Stream stream, long errorCode,
                                     ChannelPromise promise) {
if (stream.isResetSent()) {
...
}
...
```